modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItFunctionsTest.java

```diff
@@ -335,13 +336,26 @@ public void testSubstring() {
         assertQuery("SELECT SUBSTRING('text' FROM 1 FOR null)").returns(null).check();
         assertQuery("SELECT SUBSTRING('test' FROM null FOR 2)").returns(null).check();
 
-        // uncomment after https://issues.apache.org/jira/browse/IGNITE-19686 was implemented.
-        //assertQuery("select SUBSTRING(s from i for l) from (values ('abc', null, 2)) as t (s, i, l);").returns(null).check();
-
-        assertThrowsWithCause(() -> sql("SELECT SUBSTRING('abcdefg', 1, -3)"), IgniteException.class,
-                "negative substring length");
-        assertThrowsWithCause(() -> sql("SELECT SUBSTRING('abcdefg' FROM 1 FOR -1)"), IgniteException.class,
-                "negative substring length");
+        assertQuery("select SUBSTRING(s from i for l) from (values ('abc', null, 2)) as t (s, i, l);").returns(null).check();
+
+        assertQuery("SELECT substring('1234567', 2.1, 3.1);").returns("234").check();
+        assertQuery("SELECT substring('1234567', 2.1, 3);").returns("234").check();
+        assertQuery("SELECT substring('1234567', 2, 3.1);").returns("234").check();
+        assertQuery("SELECT substring('1234567', 2.1);").returns("234567").check();
```

Comment:
Maybe we should add a test case that demonstrates coercion from string to numeric, because it also work (at the moment of writing at least):

> assertQuery("SELECT SUBSTRING(1000, 1, '3')").returns("100").check();

Reply:
done